### ISSUE TITLE
torchPQ in a deep nets

### ISSUE THREAD

**Issue by Chen-Cai-OSU** [DeMoriarty/TorchPQ#5]
Hello,

Thank you very much for sharing the project. I am interested in using torchPQ inside a deep nets (implemented in pytorch) where in each forward pass, I will call torchPQ. I was wondering is this possible? 

Also, I saw  https://ai.googleblog.com/2020/07/announcing-scann-efficient-vector.html, have you tried some comparison with other methods?

Thank you!

**Comment by DeMoriarty** [comment]
Hi, thanks for using TorchPQ!

Yes, you can use TorchPQ in trained neural networks. However, backwarding through the topk search is not supported. 

Other ANNS methods / libraries such as scann, hnswlib and annoy are designed for CPU clusters, while TorchPQ and a part of FAISS are targeted towards GPUs. I didn't benchmark TorchPQ against these CPU algorithms because it will not be a fair comparison. Nearest neighbor search on single GPU will always be several times faster than a single core CPU.

I'm currently working on the next version of TorchPQ, and hopefully will release it before the next Monday. I would suggest you to wait for few days untill the next release.

**Comment by Chen-Cai-OSU** [comment]
Thank you very much for quick response!

I was wondering is it possible to support backprop through the topk search? Is backprop through KNN's top k operation is supported in pytorch? If so, I think the code can also be borrowed here.

**Comment by Chen-Cai-OSU** [comment]
just found a related paper https://proceedings.neurips.cc/paper/2020/hash/ec24a54d62ce57ba93a531b460fa8d18-Abstract.html

**Comment by DeMoriarty** [comment]
torch.topk is indeed differentiable, however it's a bruteforce search method, which means the database vectors need to have 32bit percision in order for topk to be differentiable. Product quantization on the other hand, is an approximate search algorithm, it quantizes database vectors to low percision codes and stores them to the search index for lower memory footprint and faster search speed (often times more than 16x compression), making it non-differentiable.

Can I ask you how large is the search space?

**Comment by DeMoriarty** [comment]
If the size of your vector database is less than 100,000 , I would recommend using brute force method for simplicity. 
otherwise, you may concider the following workaround:
1. store 2 different versions of the database vectors: one quantized, one with 32bit percision.
2. use IVFPQ Index to search the topk closest vectors in the quantized database (fast but not differentiable)
3. retrieve the 32bit percision version of the topk results, put them in a temporary nn.Parameter, then re-calculate the similarity scores (differentiable)
4. backprop and update the 32bit percision vectors using some variation of SGD
5. remove topk results from quantized database, quantize the updated 32bit vectors and add them back to quantized database.

**Comment by Chen-Cai-OSU** [comment]
Hello,

Thanks a lot. I think the size of database is around 250k 64 dim vector. Your proposed solution sounds very resonable but seems takes some more work. (I don't know how easy to achieve it in pytorch). I will probably try brute force first to see how it goes.

**Comment by Chen-Cai-OSU** [comment]
btw, are u aware of KeOps. Is there something worth a try for MIPS? I saw in their paper, the performance of KNN is quite promising. https://www.kernel-operations.io/

**Comment by DeMoriarty** [comment]
These links might be help:
[K-Nearest Neighbors search](https://www.kernel-operations.io/keops/_auto_benchmarks/plot_benchmark_KNN.html?highlight=topk)
[Automatic differentiation](https://www.kernel-operations.io/keops/_auto_tutorials/a_LazyTensors/plot_lazytensors_a.html?highlight=backprop#automatic-differentiation)
Looking at their benchmark results, it seems like keOps Topk performs really well on lower dimentional vectors (3d or 10d)

**Comment by Chen-Cai-OSU** [comment]
thanks! I will try keOps and let you my findings.
